moz_security/modules/heka/selprinc.lua

```diff
@@ -0,0 +1,125 @@
+-- file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+--[[
+# Heka Select Events with Principal Attributes
+
+This modules provides a generic interface to specify a list a events to match that
+contain information relevant to an action being undertaken by a principal (user). When
+an event is matched, fields are extracted per the event configuration and a normalized
+table is returned containing the relavent details from the event.
+
+## Functions
+
+### match
+
+The match function can be called from with process_event to attempt to match the event
+against the event configuration. The first event category where select_field matches
+select_match will be parsed and a table structure will be returned.
+
+For an event to successfully match, the subject_field, object_field, and sourceip_field
+must be present in the message.
+
+If the subject_map configuration option is used in an event configuration, the value
+extracted for the subject will be converted according to the subject_map table before
+being returned.
+
+The aux configuration within a given event category can be used to have the matcher return
+additional extracted message fields in the returned table, but are not specifically required
+and will be nil if they were not present in the event.
+
+*Arguments*
+- None
+
+*Return*
+- result (table, nil) - normalized table containing event fields, or nil of no match
+
+## Configuration examples
+
+```lua
+heka_selprinc = {
+    events = {
+        ssh = {
+            select_field     = "Fields[programname]",
+            select_match     = "^sshd$",
+            subject_field    = "Fields[user]",
+            object_field     = "Hostname",
+            sourceip_field   = "Fields[ssh_remote_ipaddr]",
+
+            aux = {
+                { "geocity", "Fields[ssh_remote_ipaddr_city]" },
+                { "geocountry", "Fields[ssh_remote_ipaddr_country]" }
+            }
+        },
+        awsconsole = {
+            select_field     = "Fields[eventType]",
+            select_match     = "^AwsConsoleSignIn$",
+            subject_field    = "Fields[userIdentity.userName]",
+            object_field     = "Fields[recipientAccountId]",
+            sourceip_field   = "Fields[sourceIPAddress]",
+            subject_map = {
+                ["An admin user"]   = "admin",
+                ["Commander Riker"] = "riker"
+            }
+        }
+    }
+}
+```
+--]]
+
+local string        = require "string"
+
+local module_name   = ...
+local module_cfg    = string.gsub(module_name, "%.", "_")
+local cfg           = read_config(module_cfg) or error(module_cfg .. " configuration not found")
+
+local pairs         = pairs
+local ipairs        = ipairs
+local type          = type
+local read_message  = read_message
+local assert        = assert
+
+local M = {}
+setfenv(1, M)
+
+local function cfgchk()
+    assert(type(cfg.events) == "table", "invalid selprinc events configuration")
```

Comment:
Just assert after reading the cfg var on line 75, you can also get rid of the assert import, or we may want more robust validation here instead of throwing at run time building the return table (as it may not fail for some time)?